### PR TITLE
Add support for multiline review comments

### PR DESCRIPTION
**Pull request by maximevw**
# Description

When creating a new review or adding comments to an existing review, we can currently only add single line comments.
This pull request adds support for single and multi-line comments using `line` and `start_line` attributes.

See: https://docs.github.com/en/rest/pulls/comments?apiVersion=2022-11-28#create-a-review-comment-for-a-pull-request
and https://docs.github.com/en/rest/pulls/reviews?apiVersion=2022-11-28#create-a-review-for-a-pull-request

I already prepared some tests for this and if you provide access to the `hub4j-test-org`, I will be able to finalize them.
I also successfully tested these changes on a personal project.

This will solve https://github.com/hub4j/github-api/issues/1645

# Before submitting a PR:

- [x] Changes must not break binary backwards compatibility. If you are unclear on how to make the change you think is needed while maintaining backward compatibility, [CONTRIBUTING.md](CONTRIBUTING.md) for details.
- [x] Add JavaDocs and other comments explaining the behavior. 
- [x] When adding or updating methods that fetch entities, add `@link` JavaDoc entries to the relevant documentation on https://docs.github.com/en/rest . 
- [x] Add tests that cover any added or changed code. This generally requires capturing snapshot test data. See [CONTRIBUTING.md](CONTRIBUTING.md) for details.
- [x] Run `mvn -D enable-ci clean install site` locally. If this command doesn't succeed, your change will not pass CI.
- [x] Push your changes to a branch other than `main`. You will create your PR from that branch.

# When creating a PR: 

- [x] Fill in the "Description" above with clear summary of the changes. This includes:
  - [x] If this PR fixes one or more issues, include "Fixes #<issue number>" lines for each issue. 
  - [x] Provide links to relevant documentation on https://docs.github.com/en/rest where possible. If not including links, explain why not.
- [x] All lines of new code should be covered by tests as reported by code coverage. Any lines that are not covered must have PR comments explaining why they cannot be covered. For example, "Reaching this particular exception is hard and is not a particular common scenario."
- [x] Enable "Allow edits from maintainers".

### CODE REVIEW

**Comment by maximevw** [comment]
@bitwiseman Would it be possible to give me access to hub4j-test-org in order to finalize this PR by generating test data?

**Comment by bitwiseman** [comment]
@maximevw  
Sending invite now.

**Comment by maximevw** [comment]
Hello @bitwiseman,

I updated the tests to finalize this PR. 
I also added these classes to coverage exceptions, as it was already done for `org.kohsuke.github.GHPullRequestReviewBuilder.DraftReviewComment`: 
- `org.kohsuke.github.GHPullRequestReviewBuilder.MultilineDraftReviewComment`
- `org.kohsuke.github.GHPullRequestReviewBuilder.SingleLineDraftReviewComment`

As it is the same principle, and it does not seem really relevant to write specific tests to cover simple getters in very basic POJOs, I preferred skip the coverage of these getters.

File: src/main/java/org/kohsuke/github/GHPullRequest.java [hunk 1]
```diff
@@ -557,6 +557,42 @@ public GHPullRequestReviewComment createReviewComment(String body, String sha, S
...
+            Integer startLine,
+            Integer endLine) throws IOException {
+        return root().createRequest()
+                .method("POST")
+                .with("body", body)
+                .with("commit_id", sha)
+                .with("path", path)
+                .with("start_line", endLine != null ? startLine : null)
+                .with("line", endLine != null ? endLine : startLine)
+                .withUrlPath(getApiRoute() + COMMENTS_ACTION)
```
Comment (bitwiseman) [inline src/main/java/org/kohsuke/github/GHPullRequest.java:591]: Wow, the doc page for this endpoint is a nightmare of "required but in some cases not required" statements. 😭

Comment (maximevw) [reply src/main/java/org/kohsuke/github/GHPullRequest.java:591]: Yes, it's not so easy to understand (and I can say the Gitlab API for multiline comments is really worse in terms of complexity than the Github one 😄). I tried to keep it as simple as possible in this client.

**Comment by maximevw** [comment]
Hello @bitwiseman 

Regarding the failed coverage test on the changes, as stated in my previous comment, it's only getters and it's very hard to cover these lines. I think we can ignore these warnings as the rest of the changes are covered.

> ## [Codecov](https://app.codecov.io/gh/hub4j/github-api/pull/1833?dropdown=coverage&src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=hub4j) Report
> Attention: Patch coverage is `78.78788%` with `7 lines` in your changes missing coverage. Please review.
> 
> > Project coverage is 81.04%. Comparing base [(`92112af`)](https://app.codecov.io/gh/hub4j/github-api/commit/92112af50d9840f313d2d13c97445d8e9bf2d7c3?dropdown=coverage&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=hub4j) to head [(`28a18ee`)](https://app.codecov.io/gh/hub4j/github-api/commit/28a18ee656b64ed77b7a48d6890e6beba2bac48a?dropdown=coverage&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=hub4j).

I don't know if you still expect anything else from me regarding this PR.

**Comment by bitwiseman** [comment]
@maximevw  
My apologies, but I made a broad change to the test data files resulting in a bunch of conflicts.  I've attempted a merge, but I think some of the test files still need updating. 

Re: Coverage: 
I'd prefer to test the getters.  They're not hard to do and we've had bugs show up when they weren't.  (Side note: we had a glitch in coverage reporting for the past week. I had to fix that before responding to your questions.)

Finally, with the complexity of the API, would it make more sense to have a `GHPullRequestCommentBuilder` rather than trying to model the behavior via a bunch of ternary operators in the request call? This feels like it could be a bug farm.

**Comment by maximevw** [comment]
@bitwiseman 
Thanks for your review, I pushed some changes, as requested:
* Create a `GHPullRequestReviewCommentBuilder` to avoid too complex operations in the request call method
* Add coverage for uncovered getter methods

File: src/main/java/org/kohsuke/github/GHPullRequest.java [hunk 2]
```diff
@@ -522,6 +522,15 @@ public GHPullRequestReviewBuilder createReview() {
         return new GHPullRequestReviewBuilder(this);
     }
 
+    /**
+     * Create gh pull request review comment builder.
+     *
+     * @return the gh pull request review comment builder.
+     */
+    public GHPullRequestReviewCommentBuilder createReviewComment() {
```
Comment (bitwiseman) [inline src/main/java/org/kohsuke/github/GHPullRequest.java:530]: 👍

File: src/main/java/org/kohsuke/github/GHPullRequest.java [hunk 3]
```diff
@@ -536,7 +545,9 @@ public GHPullRequestReviewBuilder createReview() {
      * @return the gh pull request review comment
      * @throws IOException
      *             the io exception
+     * @deprecated use {@link #createReviewComment()}
      */
+    @Deprecated
     public GHPullRequestReviewComment createReviewComment(String body, String sha, String path, int position)
```
Comment (bitwiseman) [inline src/main/java/org/kohsuke/github/GHPullRequest.java:551]: Replace the implementation of this method with `GHPullRequestReviewCommentBuilder`.

File: src/main/java/org/kohsuke/github/GHPullRequestReviewCommentBuilder.java [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+     * The position in the diff where you want to add a review comment.
+     *
+     * @param position
+     *            the position
+     * @return the gh pull request review comment builder
+     * @deprecated This parameter is deprecated in GitHub API, use {@link #line(int)} instead.
+     */
+    @Deprecated
+    public GHPullRequestReviewCommentBuilder position(int position) {
+        builder.with("position", position);
```
Comment (bitwiseman) [inline src/main/java/org/kohsuke/github/GHPullRequestReviewCommentBuilder.java:75]: Let's make this internal at most.  If it is not what people should use, we should not let them use it at all.

File: src/main/java/org/kohsuke/github/GHPullRequestReviewCommentBuilder.java [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    }
+
+    /**
+     * The first line in the pull request diff that your multi-line comment applies to.
+     *
+     * @param line
+     *            the line number
+     * @return the gh pull request review comment builder
+     */
+    public GHPullRequestReviewCommentBuilder startLine(int line) {
```
Comment (bitwiseman) [inline src/main/java/org/kohsuke/github/GHPullRequestReviewCommentBuilder.java:99]: How about `lines(int startLine, int endLine)`?  Have it set both properties and have `line()` set `line` and clear `start_line`.  

And update the javadoc to indicate the `line()` and `lines()` overwrite each other.  Make sense?

File: src/main/java/org/kohsuke/github/GHPullRequestReviewBuilder.java [hunk 6]
```diff
@@ -75,19 +75,53 @@ public GHPullRequestReviewBuilder event(GHPullRequestReviewEvent event) {
...
      * @param path
+     *            The relative path to the file that necessitates a review comment.
+     * @param position
      *            The position in the diff where you want to add a review comment. Note this value is not the same as
      *            the line number in the file. For help finding the position value, read the note below.
-     * @param position
-     *            Text of the review comment.
      * @return the gh pull request review builder
      */
     public GHPullRequestReviewBuilder comment(String body, String path, int position) {
```
Comment (bitwiseman) [inline src/main/java/org/kohsuke/github/GHPullRequestReviewBuilder.java:86]: Deprecated?

File: src/main/java/org/kohsuke/github/GHPullRequestReviewBuilder.java [hunk 7]
```diff
@@ -75,19 +75,53 @@ public GHPullRequestReviewBuilder event(GHPullRequestReviewEvent event) {
...
+     *            Text of the review comment.
+     * @param path
+     *            The relative path to the file that necessitates a review comment.
+     * @param startLine
+     *            The first line in the pull request diff that the multi-line comment applies to.
+     * @param line
+     *            The last line of the range that the comment applies to.
+     * @return the gh pull request review builder
+     */
+    public GHPullRequestReviewBuilder multiLineComment(String body, String path, int startLine, int line) {
```
Comment (bitwiseman) [inline src/main/java/org/kohsuke/github/GHPullRequestReviewBuilder.java:104]:
```suggestion
    public GHPullRequestReviewBuilder multiLineComment(String body, String path, int startLine, int endLine) {
```

**Review by bitwiseman — CHANGES_REQUESTED** [review]
Suggestions and questions.  Open to discussion.

File: src/main/java/org/kohsuke/github/GHPullRequestReviewBuilder.java (hunk 6 above)
Comment (maximevw) [reply src/main/java/org/kohsuke/github/GHPullRequestReviewBuilder.java:86]: In the documentation of the GitHub API (https://docs.github.com/en/rest/pulls/reviews?apiVersion=2022-11-28#create-a-review-for-a-pull-request), the use of position when adding a comment at the moment we create a review is not marked as deprecated (surprisingly), that's why I kept this not depreacted.

The position parameter is only marked deprecated when we create a comment on a pull request review (https://docs.github.com/en/rest/pulls/comments?apiVersion=2022-11-28#create-a-review-comment-for-a-pull-request). In doubt, I just followed the current documentation.

**Comment by maximevw** [comment]
@bitwiseman I pushed the requested modifications and answered to your question regarding the deprecation of `GHPullRequestReviewBuilder.comment(String, String, int)`.

File: src/main/java/org/kohsuke/github/GHPullRequestReviewCommentBuilder.java [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+     * Be careful, this value can be overridden by {@link #lines(int, int)}.
+     * </p>
+     *
+     * @param line
+     *            the line number
+     * @return the gh pull request review comment builder
+     */
+    public GHPullRequestReviewCommentBuilder line(int line) {
+        builder.with("line", line);
+        return this;
```
Comment (bitwiseman) [inline src/main/java/org/kohsuke/github/GHPullRequestReviewCommentBuilder.java:93]:
```suggestion
        builder.with("line", line);
        builder.remove("start_line");
        return this;
```

File: src/main/java/org/kohsuke/github/GHPullRequestReviewCommentBuilder.java [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+    }
+
+    /**
+     * For a single line comment, the line of the blob in the pull request diff that the comment applies to. For a
+     * multi-line comment, the last line of the range that your comment applies to (in this case, it's recommended to
+     * use {@link #lines(int, int)} instead).
+     * <p>
+     * Be careful, this value can be overridden by {@link #lines(int, int)}.
+     * </p>
```
Comment (bitwiseman) [inline src/main/java/org/kohsuke/github/GHPullRequestReviewCommentBuilder.java:85]:
```suggestion
     *  A single line of the blob in the pull request diff that the comment applies to.
     * <p>
     * {@link #line(int)} and {@link #lines(int, int)} will overwrite each other's values.
     * </p>
```

File: src/main/java/org/kohsuke/github/GHPullRequestReviewCommentBuilder.java (hunk 8 above)
Comment (bitwiseman) [inline src/main/java/org/kohsuke/github/GHPullRequestReviewCommentBuilder.java:93]:
```suggestion
        builder.with("line", line);
        builder.remove("start_line");
        return this;
```

File: src/main/java/org/kohsuke/github/GHPullRequestReviewCommentBuilder.java [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+     * <p>
+     * Be careful, the end line value can be overridden by {@link #line(int)} (usually used for single line comments).
+     * </p>
+     *
+     * @param startLine
+     *            the start line number of the comment
+     * @param endLine
+     *            the end line number of the comment
+     * @return the gh pull request review comment builder
+     */
```
Comment (bitwiseman) [inline src/main/java/org/kohsuke/github/GHPullRequestReviewCommentBuilder.java:107]:
```suggestion
    /**
     * The range of lines in the pull request diff that this comment applies to.
     * <p>
     * {@link #line(int)} and {@link #lines(int, int)} will overwrite each other's values.
     * </p>
     *
     * @param startLine
     *            the start line number of the comment
     * @param endLine
     *            the end line number of the comment
     * @return the gh pull request review comment builder
     */
```

**Comment by maximevw** [comment]
Thanks @bitwiseman for the last modifications, I was a little hasty when I pushed my changes... 😕

File: src/main/java/org/kohsuke/github/GHPullRequestReviewBuilder.java (hunk 6 above)
Comment (maximevw) [reply src/main/java/org/kohsuke/github/GHPullRequestReviewBuilder.java:86]: @bitwiseman Do you still want to mark this function as deprecated or leave it as is for now? (despite the confusing GH documentation on this point...)

File: src/main/java/org/kohsuke/github/GHPullRequest.java [hunk 11]
```diff
@@ -536,18 +545,12 @@ public GHPullRequestReviewBuilder createReview() {
...
-        return root().createRequest()
-                .method("POST")
-                .with("body", body)
-                .with("commit_id", sha)
-                .with("path", path)
-                .with("position", position)
-                .withUrlPath(getApiRoute() + COMMENTS_ACTION)
-                .fetch(GHPullRequestReviewComment.class)
-                .wrapUp(this);
+        return createReviewComment().body(body).commitId(sha).path(path).position(position).create();
```
Comment (bitwiseman) [inline src/main/java/org/kohsuke/github/GHPullRequest.java:553]: Looks like your test changes removed coverage for this.

File: src/test/java/org/kohsuke/github/GHPullRequestTest.java [hunk 12]
```diff
@@ -268,11 +319,29 @@ public void pullRequestReviewComments() throws Exception {
...
-            // System.out.println(p.getUrl());
             assertThat(p.listReviewComments().toList(), is(empty()));
-            p.createReviewComment("Sample review comment", p.getHead().getSha(), "README.md", 1);
+
+            p.createReviewComment()
+                    .commitId(p.getHead().getSha())
+                    .body("Sample review comment")
+                    .path("README.md")
+                    .position(1)
+                    .create();
```
Comment (bitwiseman) [inline src/test/java/org/kohsuke/github/GHPullRequestTest.java:329]:
```suggestion

             // Call the deprecated method to ensure continued support
             p.createReviewComment("Sample review comment", p.getHead().getSha(), "README.md", 1);
```

File: src/main/java/org/kohsuke/github/GHPullRequestReviewBuilder.java (hunk 6 above)
Comment (bitwiseman) [reply src/main/java/org/kohsuke/github/GHPullRequestReviewBuilder.java:86]: Nah, leave it.

File: src/test/java/org/kohsuke/github/GHPullRequestTest.java [hunk 13]
```diff
@@ -268,11 +319,25 @@ public void pullRequestReviewComments() throws Exception {
         String name = "pullRequestReviewComments";
         GHPullRequest p = getRepository().createPullRequest(name, "test/stable", "main", "## test");
         try {
-            // System.out.println(p.getUrl());
             assertThat(p.listReviewComments().toList(), is(empty()));
-            p.createReviewComment("Sample review comment", p.getHead().getSha(), "README.md", 1);
+
+             // Call the deprecated method to ensure continued support
+             p.createReviewComment("Sample review comment", p.getHead().getSha(), "README.md", 1);
```
Comment (bitwiseman) [inline src/test/java/org/kohsuke/github/GHPullRequestTest.java:325]:
```suggestion
            // Call the deprecated method to ensure continued support
            p.createReviewComment("Sample review comment", p.getHead().getSha(), "README.md", 1);
```

**Review by bitwiseman — APPROVED** [review]

**Pull request merged**